### PR TITLE
Adds API info to RFC

### PR DESCRIPTION
**Pull request by elorzafe**

### CODE REVIEW

File: RFC.md [hunk 1]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
 
 Try out the new `storage` types here: https://www.typescriptlang.org/play#gist/292f4e24178bfca5881aa20961b930dc
 
-# `API` & `Datastore` Changes
+# `API` category Changes
```
Comment (jimblanc) [inline RFC.md]:
```suggestion
# `API` Category Changes
Amplify is proposing the following changes for the `API` category.
```

(for top level sections)

File: RFC.md [hunk 2]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+
+API.get(apiName, path, myInit)
+  .then((response) => {
+    // Add your code here
+  })
+  .catch((error) => {
+    console.log(error.response);
+  });
+```
+
```
Comment (jimblanc) [inline RFC.md]:
```suggestion
**Proposed Amplify v6 (`aws-amplify@6`)**
```

File: RFC.md [hunk 3]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+```typescript
+API.get({
+    apiName: 'MyApi',
+    path: '/items',
+    authMode: 'AWS_IAM'
+}, myInit)
+  .then(result => {
```
Comment (jimblanc) [inline RFC.md]:
```suggestion
  .then((result) => {
```

Inconsistent parens

File: RFC.md [hunk 4]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+    // Add your code here
+  }).catch((error) => {
+    console.log(error.response);
+  });
+```
+
+## Adding TypeScript generics to request body and request response
```
Comment (jimblanc) [inline RFC.md]:
```suggestion
## Adding TypeScript generics to request body and response
To improve developer experience and permit more strict typing we will be adding generic support to our `API` category APIs.
```

File: RFC.md [hunk 5]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+  .catch((error) => {
+    console.log(error.response);
+  });
+```
+
+```typescript
+API.get({
+    apiName: 'MyApi',
+    path: '/items',
+    authMode: 'AWS_IAM'
```
Comment (jimblanc) [inline RFC.md]:
```suggestion
  apiName: 'MyApi',
  path: '/items',
  authMode: 'AWS_IAM'
```

File: RFC.md [hunk 6]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+    console.log(error.response);
+  });
+```
+
+## Adding TypeScript generics to request body and request response
+
+**Amplify v5 (`aws-amplify@5`)**
+
+Not possible to use generics for request body or response body
+
```
Comment (jimblanc) [inline RFC.md]:
```suggestion
Amplify v5 does not support using generics for the request body or response.
```

File: RFC.md [hunk 7]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
-# `API` & `Datastore` Changes
+# `API` category Changes
+
+## First param is an object with named parameters
+
+**Amplify v5 (`aws-amplify@5`)**
+
+```typescript
+const apiName = 'MyApiName';
+const path = '/path';
```
Comment (jimblanc) [inline RFC.md:353]: Nit: IMO these should just be inline (e.g. `API.get('MyApiName', '/path', ...)`)

File: RFC.md [hunk 8]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+    headers: {
+        "Content-type": "text/plain",
+    },
+    body: "this is my content"
+}).then(result => {
+    result.body.data.forEach(value => console.log(value));
+});
+```
+
+## GraphQL operations are splitted between query, mutation and subscription
```
Comment (jimblanc) [inline RFC.md]:
```suggestion
## GraphQL operations have been split into query, mutation, and subscription
To better capture customer intent and simplify API types we will split up the `graphql` API into individual APIs for queries, mutations, and subscriptions.
```

File: RFC.md [hunk 9]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+
+**Amplify v5 (`aws-amplify@5`)**
+
+```typescript
+import { API } from "aws-amplify";
+import * as mutations from './graphql/mutations';
+import { GraphQLQuery, GraphQLSubscription } from '@aws-amplify/api';
+import { CreateTodoInput, CreateTodoMutation, OnCreateTodoSubscription } from './API';
+import * as subscriptions from './graphql/subscriptions';
+
```
Comment (jimblanc) [inline RFC.md:428]:
```suggestion
```

IMO we should omit imports from samples.

File: RFC.md [hunk 10]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+  name: 'Todo 1',
+  description: 'Learn AWS AppSync'
+};
+
+const newTodo = await API.graphql<GraphQLQuery<CreateTodoMutation>>({ 
+  query: mutations.createTodo, 
+  variables: { input: todoDetails }
+});
+
+const sub = API.graphql<GraphQLSubscription<OnCreateTodoSubscription>>(
```
Comment (jimblanc) [inline RFC.md]:
```suggestion
const subscription = API.graphql<GraphQLSubscription<OnCreateTodoSubscription>>(
```

File: RFC.md [hunk 11]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+    apiName: '',
+    path: '/',
+    authMode: "API_KEY"
+}, {
+    headers: {
+        "Content-type": "text/plain",
+    },
+    body: "this is my content"
+}).then(result => {
+    result.body.data.forEach(value => console.log(value));
```
Comment (jimblanc) [inline RFC.md]:
```suggestion
}).then((result) => {
    result.body.data.forEach((value) => console.log(value));
```

(consistent parens)

File: RFC.md [hunk 12]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+```
+**Proposed Amplify v6 (`aws-amplify@6`)**
+
+```typescript
+type MyQueryType = {
+    result: {
+        id: string,
+        name: string,
+        description: string
+    }
```
Comment (jimblanc) [inline RFC.md]:
```suggestion
  result: {
    id: string,
    name: string,
    description: string
  }
```

File: RFC.md [hunk 13]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+}
+
+API.graphqlQuery<MyQueryType>({
+    document: `query getTodo...`,
```
Comment (jimblanc) [inline RFC.md]:
```suggestion
  document: `query getTodo...`,
```

File: RFC.md [hunk 14]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+```typescript
+type MyQueryType = {
+    result: {
+        id: string,
+        name: string,
+        description: string
+    }
+}
+
+API.graphqlQuery<MyQueryType>({
```
Comment (jimblanc) [inline RFC.md:457]: It seems like the "query" example is missing from v5 sample above, we should make sure the v5 & v6 samples capture identical tasks to make comparison easier.

File: RFC.md [hunk 15]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+    document: `query getTodo...`,
+})
+.then(result => {
+    console.log(`Todo : ${result.data?.id}: ${result.data?.name} (${result.data?.description})`);
```
Comment (jimblanc) [inline RFC.md]:
```suggestion
.then((result) => {
  console.log(`Todo : ${result.data?.id}: ${result.data?.name} (${result.data?.description})`);
```

File: RFC.md [hunk 16]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+    variables: {
+        id: number,
+        name: string,
+        description: string
+    },
+    result: {
+      id: number, 
+      name: string, 
+      description: string
+    }
```
Comment (jimblanc) [inline RFC.md]:
```suggestion
  variables: {
    id: number,
    name: string,
    description: string
  },
  result: {
    id: number, 
    name: string, 
    description: string
  }
```

File: RFC.md [hunk 17]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+}
+
+API.graphqlMutation<MyMutationType>({
+    document: `mutation createTodo....`,
+    variables: {
+        id: 123,
+        name: 'My Todo',
+        description: 'This is a todo'
+    }
```
Comment (jimblanc) [inline RFC.md]:
```suggestion
  document: `mutation createTodo....`,
  variables: {
    id: 123,
    name: 'My Todo',
    description: 'This is a todo'
  }
```

File: RFC.md (hunk 13 above)
Comment (jimblanc) [inline RFC.md]: Any reason to not using the codegen'd docs here? Don't want to imply that customers will need to write their own documents.

File: RFC.md [hunk 18]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+    },
+    result: {
+      id: number, 
+      name: string, 
+      description: string
+    }
+}
+
+API.graphqlMutation<MyMutationType>({
+    document: `mutation createTodo....`,
```
Comment (jimblanc) [inline RFC.md]: Ditto about using existing document.

File: RFC.md [hunk 19]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+    variables: {
+        id: 123,
+        name: 'My Todo',
+        description: 'This is a todo'
+    }
+})
+.then(result => {
+    console.log(`Todo : ${result.data?.id}: ${result.data?.name} (${result.data?.description})`);
```
Comment (jimblanc) [inline RFC.md]:
```suggestion
.then((result) => {
  console.log(`Todo : ${result.data?.id}: ${result.data?.name} (${result.data?.description})`);
```

File: RFC.md [hunk 20]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+});
+
+API.graphqlSubscription<MyQueryType>({ document: `subscription OnCreateTodo...`})
```
Comment (jimblanc) [inline RFC.md:489]: Ditto about using document

File: RFC.md [hunk 21]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+.subscribe({
+    next: (result) => console.log(`Todo info: ${result.data?.id}: ${result.data?.name} (${result.data?.description})`),
```
Comment (jimblanc) [inline RFC.md]:
```suggestion
  next: (result) => console.log(`Todo info: ${result.data?.id}: ${result.data?.name} (${result.data?.description})`),
```

File: RFC.md [hunk 22]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+});
+```
+
+## Errors can be narrowed down 
+
+**Amplify v5 (`aws-amplify@5`)**
+Errors cannot be narrowed down
```
Comment (jimblanc) [inline RFC.md]:
```suggestion
Amplify v5 does not support narrowing down errors.
```

File: RFC.md [hunk 23]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+
+**Proposed Amplify v6 (`aws-amplify@6`)**
+
+```typescript
+import { HTTPError, NetworkError, BlockedError, CancelledError}
+
```
Comment (jimblanc) [inline RFC.md]:
```suggestion
```

File: RFC.md [hunk 24]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+API.get({
+    apiName: 'myApi',
+    path: '/'
```
Comment (jimblanc) [inline RFC.md]:
```suggestion
  apiName: 'myApi',
  path: '/'
```

File: RFC.md [hunk 25]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+}).then(result => {
+    // do something with result
```
Comment (jimblanc) [inline RFC.md]:
```suggestion
}).then((result) => {
  // do something with result
```

File: RFC.md [hunk 26]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+        // Maybe I want to retry
+    } else if (err instanceof HTTPError) {
+        // Check if something is going with my request parameters
+    } else if (err instanceof CancelledError) {
+        // this is fine I trigger the cancellation
+    } else if (err instanceof BlockedError) {
+        // This is CORS stuff
+    } else {
+        // other error
+    }
```
Comment (jimblanc) [inline RFC.md]:
```suggestion
  if (err instanceof NetworkError) {
    // Consider retrying
  } else if (err instanceof HTTPError) {
    // Check request parameters for mistakes
  } else if (err instanceof CancelledError) {
    // Request was cancelled
  } else if (err instanceof BlockedError) {
    // CORS related error
  } else {
    // Other error
  }
```

File: RFC.md [hunk 27]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+});
+```
 
 Try out the new types here: TODO Playground Link
```
Comment (jimblanc) [inline RFC.md]: Could you get this updated? Storage example here: https://www.typescriptlang.org/play#gist/292f4e24178bfca5881aa20961b930dc

File: RFC.md [hunk 28]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
...
+    console.log(error.response);
+  });
+```
+
+```typescript
+API.get({
+    apiName: 'MyApi',
+    path: '/items',
+    authMode: 'AWS_IAM'
+}, myInit)
```
Comment (jimblanc) [inline RFC.md:376]: Doesn't need to happen now, but why was `myInit` excluded from the param object? We should try to be consistent between category APIs one way or the other.

File: RFC.md [hunk 29]
```diff
@@ -341,6 +341,183 @@ const uploadedObjectReference = await uploadTask.result;
 
 Try out the new `storage` types here: https://www.typescriptlang.org/play#gist/292f4e24178bfca5881aa20961b930dc
 
-# `API` & `Datastore` Changes
+# `API` category Changes
+
+## First param is an object with named parameters
+
```
Comment (jimblanc) [inline RFC.md]:
```suggestion
To improve the readability of our APIs we will be introducing an object parameter to capture request parameters.
```

File: RFC.md (hunk 7 above)
Comment (elorzafe) [reply RFC.md:353]: I copy/pasted this from our docs actually. Is just to show is better the name approach so is easy to do it inline

File: RFC.md (hunk 28 above)
Comment (elorzafe) [reply RFC.md:376]: You mean having two arguments?

File: RFC.md (hunk 9 above)
Comment (elorzafe) [reply RFC.md:428]: Yeah, it was to make explicit the thing is annoying lol

File: RFC.md (hunk 13 above)
Comment (elorzafe) [reply RFC.md]: Good call, it was more about telling the customer they could also use their own queries (not depending on CLI) I didnt like the idea of having too many things that they might ask where is this coming from

File: RFC.md (hunk 27 above)
Comment (elorzafe) [reply RFC.md]: I miss that, I thought it was there, updating now

File: RFC.md (hunk 28 above)
Comment (jimblanc) [reply RFC.md:376]: Yeah in this PR it has two arguments but I was wondering if it should just be one object param

**Pull request merged**